node/pkg/ecdsasigner/cloudkms.go

```diff
@@ -0,0 +1,166 @@
+	}
+
+	pubKey := ethcrypto.PubkeyToAddress(*c.publicKey)
+
+	sig, err := parseSignature(signResult.Signature, digest, pubKey)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to parse KMS signature: %w", err)
+	}
+
+	return sig, nil
+}
+
+func (c KMSClient) PublicKey() ecdsa.PublicKey {
+	return *c.publicKey
+}
+
+// Private Functions
+func getKMSPublicKey(
+	ctx context.Context,
+	keyId string,
+	client kms.KeyManagementClient,
+) (*ecdsa.PublicKey, error) {
+	request := &kmspb.GetPublicKeyRequest{Name: keyId}
+
+	kmsPublicKey, err := client.GetPublicKey(ctx, request)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to fetch public key from the KMS API: %v", err)
+	}
+
+	pemBlock, rest := pem.Decode([]byte(kmsPublicKey.Pem))
+	if len(rest) > 0 {
+		return nil, fmt.Errorf("PEM block contains more than just public key")
+	}
+
+	var pki struct {
+		Raw       asn1.RawContent
+		Algorithm pkix.AlgorithmIdentifier
+		PublicKey asn1.BitString
+	}
+	if rest, err := asn1.Unmarshal(pemBlock.Bytes, &pki); err != nil {
+		return nil, fmt.Errorf("Failed to unmarshal KMS public key from pem block: %v", err)
+	} else if len(rest) != 0 {
+		return nil, fmt.Errorf("Trailing data after ASN.1 unmarshalling of the KMS public key")
+	}
+
+	asn1Data := pki.PublicKey.RightAlign()
+	if asn1Data[0] != 4 {
+		return nil, fmt.Errorf("Only uncompressed public key is supported")
+	}
+
+	secp256k1Curve := ethcrypto.S256()
+	p := secp256k1Curve.Params().P
+	plen := bitsToBytes(p.BitLen())
+
+	asn1PubKeyBytes := asn1Data[1:]
+	if len(asn1PubKeyBytes) != 2*plen {
+		return nil, fmt.Errorf("Public key has incorrect size, got %d, expects %d", len(asn1PubKeyBytes), 2*plen)
+	}
+
+	var x, y big.Int
+	x.SetBytes(asn1PubKeyBytes[:plen])
+	y.SetBytes(asn1PubKeyBytes[plen:])
+
+	if x.Cmp(p) >= 0 || y.Cmp(p) >= 0 || !secp256k1Curve.IsOnCurve(&x, &y) {
+		return nil, fmt.Errorf("kms input is not on curve")
+	}
+
+	return &ecdsa.PublicKey{Curve: secp256k1Curve, X: &x, Y: &y}, nil
+}
+
+func parseSignature(kmsSignature []byte, digest []byte, pubKey ethcommon.Address) ([]byte, error) {
+	var parsedSig struct{ R, S *big.Int }
+	if _, err := asn1.Unmarshal(kmsSignature, &parsedSig); err != nil {
+		return nil, fmt.Errorf("Fail to unmarshal KMS signature: %w", err)
+	}
+
+	curveOrderLen := 32
+	signature := make([]byte, 2*curveOrderLen)
+
+	// left pad R and S with zeroes
+	rBytes := parsedSig.R.Bytes()
+	sBytes := parsedSig.S.Bytes()
+	copy(signature[curveOrderLen-len(rBytes):], rBytes)
+	copy(signature[len(signature)-len(sBytes):], sBytes)
+
+	return appendV(signature, digest, pubKey)
+}
+
+func appendV(sig []byte, digest []byte, pubKey ethcommon.Address) ([]byte, error) {
+	sigWithV := append(sig, 0)
+
+	for i := 0; i < 4; i++ {
+		pk, err := ethcrypto.Ecrecover(digest, sigWithV)
+		signer_pk := ethcommon.BytesToAddress(ethcrypto.Keccak256(pk[1:])[12:])
+		if err == nil && signer_pk == pubKey {
+			return sigWithV, nil
+		}
+		sigWithV = append(sig, byte(i))
+	}
```

Comment:
The loop block code will execute twice for `sigWithV = append(sig, 0)` if the `V` is not 0.

```suggestion
	for i := 0; i < 4; i++ {
	        sigWithV := append(sig, byte(i))
		pk, err := ethcrypto.Ecrecover(digest, sigWithV)
		signer_pk := ethcommon.BytesToAddress(ethcrypto.Keccak256(pk[1:])[12:])
		if err == nil && signer_pk == pubKey {
			return sigWithV, nil
		}
	}
```